File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/azure/AzureBlobStorageStreamCopierFactory.java
/*
 * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.jdbc.copy.azure;

import com.azure.storage.blob.specialized.SpecializedBlobClientBuilder;
import io.airbyte.db.jdbc.JdbcDatabase;
import io.airbyte.integrations.destination.ExtendedNameTransformer;
import io.airbyte.integrations.destination.jdbc.SqlOperations;
import io.airbyte.integrations.destination.jdbc.copy.StreamCopier;
import io.airbyte.integrations.destination.jdbc.copy.StreamCopierFactory;
import io.airbyte.protocol.models.v0.AirbyteStream;
import io.airbyte.protocol.models.v0.ConfiguredAirbyteStream;
import io.airbyte.protocol.models.v0.DestinationSyncMode;

public abstract class AzureBlobStorageStreamCopierFactory implements StreamCopierFactory<AzureBlobStorageConfig> {

  @Override
  public StreamCopier create(String configuredSchema,
                             AzureBlobStorageConfig azureBlobConfig,
                             String stagingFolder,
                             ConfiguredAirbyteStream configuredStream,
                             ExtendedNameTransformer nameTransformer,
                             JdbcDatabase db,
                             SqlOperations sqlOperations) {
    try {
      AirbyteStream stream = configuredStream.getStream();
      DestinationSyncMode syncMode = configuredStream.getDestinationSyncMode();
      String schema = StreamCopierFactory.getSchema(stream.getNamespace(), configuredSchema, nameTransformer);
      String streamName = stream.getName();

      final SpecializedBlobClientBuilder specializedBlobClientBuilder = new SpecializedBlobClientBuilder()
          .endpoint(azureBlobConfig.getEndpointUrl())
          .sasToken(azureBlobConfig.getSasToken())
          .containerName(azureBlobConfig.getContainerName());

      return create(stagingFolder, syncMode, schema, streamName, specializedBlobClientBuilder, db, azureBlobConfig, nameTransformer, sqlOperations);
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
  }

  public abstract StreamCopier create(String stagingFolder,
                                      DestinationSyncMode syncMode,
                                      String schema,
                                      String streamName,
                                      SpecializedBlobClientBuilder specializedBlobClientBuilder,
                                      JdbcDatabase db,
                                      AzureBlobStorageConfig azureBlobConfig,
                                      ExtendedNameTransformer nameTransformer,
                                      SqlOperations sqlOperations)
      throws Exception;

}

File: airbyte-cdk/java/airbyte-cdk/azure-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/jdbc/copy/azure/AzureBlobStorageStreamCopierFactory.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.destination.jdbc.copy.azure

import com.azure.storage.blob.specialized.SpecializedBlobClientBuilder
import io.airbyte.cdk.db.jdbc.JdbcDatabase
import io.airbyte.cdk.integrations.destination.StandardNameTransformer
import io.airbyte.cdk.integrations.destination.jdbc.SqlOperations
import io.airbyte.cdk.integrations.destination.jdbc.copy.StreamCopier
import io.airbyte.cdk.integrations.destination.jdbc.copy.StreamCopierFactory
import io.airbyte.cdk.integrations.destination.jdbc.copy.StreamCopierFactory.Companion.getSchema
import io.airbyte.protocol.models.v0.ConfiguredAirbyteStream
import io.airbyte.protocol.models.v0.DestinationSyncMode

abstract class AzureBlobStorageStreamCopierFactory : StreamCopierFactory<AzureBlobStorageConfig> {
    override fun create(
        configuredSchema: String?,
        config: AzureBlobStorageConfig,
        stagingFolder: String?,
        configuredStream: ConfiguredAirbyteStream?,
        nameTransformer: StandardNameTransformer?,
        db: JdbcDatabase?,
        sqlOperations: SqlOperations?
    ): StreamCopier {
        try {
            val stream = configuredStream!!.stream
            val syncMode = configuredStream.destinationSyncMode
            val schema = getSchema(stream.namespace, configuredSchema!!, nameTransformer!!)
            val streamName = stream.name

            val specializedBlobClientBuilder =
                SpecializedBlobClientBuilder()
                    .endpoint(config.endpointUrl)
                    .sasToken(config.sasToken)
                    .containerName(config.containerName)

            return create(
                stagingFolder,
                syncMode,
                schema,
                streamName,
                specializedBlobClientBuilder,
                db,
                config,
                nameTransformer,
                sqlOperations
            )
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }

    @Throws(Exception::class)
    abstract fun create(
        stagingFolder: String?,
        syncMode: DestinationSyncMode?,
        schema: String?,
        streamName: String?,
        specializedBlobClientBuilder: SpecializedBlobClientBuilder?,
        db: JdbcDatabase?,
        azureBlobConfig: AzureBlobStorageConfig?,
        nameTransformer: StandardNameTransformer?,
        sqlOperations: SqlOperations?
    ): StreamCopier
}
